src/validate.rs

```diff
@@ -1,33 +1,1490 @@
+) -> Result<StringSignature, Vec<ValidationError>> {
+    let mut errors = Vec::new();
+    let mut signature = StringSignature {
+        parameters: HashMap::new(),
+        nonpositional_count: BTreeMap::new(),
+    };
+
+    let mut pos = 0;
+    for fragment in &base.fragments {
+        if let FragmentContent::Command(cmd) = &fragment.content {
+            if let Some(info) = COMMANDS
+                .into_iter()
+                .find(|ci| ci.name == cmd.name && ci.dialects.contains(&dialect))
+            {
+                if info.parameters.is_empty() {
+                    if let Some(index) = cmd.index {
+                        errors.push(ValidationError {
+                            critical: true,
+                            position: Some(fragment.position),
+                            message: format!(
+                                "Command '{{{}}}' cannot have a position reference.",
+                                cmd.name
+                            ),
+                            suggestion: Some(format!("Remove '{}:'.", index)),
+                        });
+                    }
+                    let norm_name = String::from(info.get_norm_name());
+                    if let Some(existing) = signature.nonpositional_count.get_mut(&norm_name) {
+                        existing.1 += 1;
+                    } else {
+                        signature
+                            .nonpositional_count
+                            .insert(norm_name, (info.occurence.clone(), 1));
+                    }
+                } else {
+                    if let Some(index) = cmd.index {
+                        pos = index;
+                    }
+                    if let Some(existing) = signature.parameters.insert(pos, info) {
+                        errors.push(ValidationError {
+                            critical: true,
+                            position: Some(fragment.position),
+                            message: format!(
+                                "Command '{{{}:{}}}' references the same position as '{{{}:{}}}' before.",
+                                pos, cmd.name, pos, existing.name
+                            ),
+                            suggestion: Some(String::from("Assign unique position references.")),
+                        });
+                    }
+                    pos += 1;
+                }
+            } else {
+                errors.push(ValidationError {
+                    critical: true,
+                    position: Some(fragment.position),
+                    message: format!("Unknown string command '{{{}}}'.", cmd.name),
+                    suggestion: None,
+                });
+            }
+        }
+    }
+
+    if errors.is_empty() {
+        Ok(signature)
+    } else {
+        Err(errors)
+    }
+}
+
+fn validate_string(
+    config: &LanguageConfig,
+    test: &ParsedString,
+    base: Option<&ParsedString>,
+) -> Vec<ValidationError> {
```

Comment:
This function is over 300 lines long; any chance we can split it up a bit?

Reply:
Yes, I am aware. There is also some code duplication with other methods. I need some kind of generator/iterator to loop over commands and figureing out positional parameter references.
Also part of "bells and whistles"